src/v/utils/hybrid_clock.h

```diff
@@ -0,0 +1,187 @@
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+#include "base/likely.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/manual_clock.hh>
+
+#include <chrono>
+
+/// The hybrid_clock is a thin wrapper around the seastar::lowres_clock
+/// that can be swapped from seastar::manual_clock in tests by flipping
+/// a runtime flag (per shard). The timestamps and durations are using
+/// last bit as a flag indicating the type of clock used to produce
+/// them.
+class hybrid_clock {
+    struct duration_with_flag {
+        uint64_t nanoseconds{0};
+        bool manual{false};
+    };
+
+public:
+    // Same as std::chrono::steady_clock::rep
+    //
+    using duration = std::chrono::nanoseconds;
+    using rep = duration::rep;
+    using period = duration::period;
+    using time_point = std::chrono::time_point<hybrid_clock, duration>;
+    static constexpr bool is_steady = true;
+
+    inline static time_point now() noexcept;
+
+    inline static void update() noexcept;
+
+    inline static void advance(duration d);
+
+    inline static void set_manual_mode(bool mode);
+
+    inline static bool is_manual(const time_point& t);
+
+    friend bool operator==(const time_point& lhs, const time_point& rhs);
+    friend auto operator<=>(const time_point& lhs, const time_point& rhs);
+    friend duration operator-(const time_point& lhs, const time_point& rhs);
+    friend time_point operator-(const time_point& lhs, const duration& rhs);
+    friend time_point operator+(const time_point& lhs, const duration& rhs);
+
+private:
+    inline static void throw_clock_mismatch() {
+        throw std::logic_error("clock mismatch");
+    }
+    // This will set zero bit of the duration to 0 if the lowres_clock  is used
+    // or to 1 if manual clock is used. That's totally fine given that the clock
+    // is lowres and the duration is expressed in nanoseconds. So the bit is
+    // normally set to zero and will only be flipped to 1 in case if manual
+    // clock is used.
+    inline static void duration_bit_hack(duration& target, bool manual);
+    // Extract the actual timestamp with the flag
+    inline static duration_with_flag decode(const duration& d);
+
+    inline static thread_local bool _manual{false}; // NOLINT
+    inline static constexpr uint64_t mask = uint64_t(0) - 2;
+};
+
+inline hybrid_clock::time_point hybrid_clock::now() noexcept {
+    auto dur = std::chrono::duration_cast<duration>(
+      seastar::lowres_clock::now().time_since_epoch());
+    if (unlikely(_manual)) {
+        auto now = seastar::manual_clock::now();
+        dur = std::chrono::duration_cast<duration>(now.time_since_epoch());
+    }
+    duration_bit_hack(dur, _manual);
+    return time_point(dur);
+}
+
+inline void hybrid_clock::update() noexcept { seastar::lowres_clock::update(); }
+
+inline void hybrid_clock::advance(hybrid_clock::duration d) {
+    seastar::manual_clock::advance(
```

Comment:
Is there potentially a good reason we should enforce that `_manual == true` here?